Add support for the tuple data type (#16776)
vim/vim: runtime/doc/eval.txt

List		An ordered sequence of items, see |List| for details.
		Example: [1, 2, ['a', 'b']]

> Tuple		An ordered immutable sequence of items, see |Tuple| for details.

Reviewer: ```suggestion
Tuple		An ordered immutable sequence of items, see |Tuple| for
		details.
```

Author: I have updated the help text to incorporate this comment.

		*E611* *E745* *E728* *E703* *E729* *E730* *E731* *E908* *E910*
		*E913* *E974* *E975* *E976* *E1319* *E1320* *E1321* *E1322*
		*E1323* *E1324* *E1520* *E1522*

> |List|, |Tuple|, |Dictionary|, |Funcref|, |Job|, |Channel|, |Blob|, |Class| and
> |object| types are not automatically converted.

Reviewer: ```suggestion
|List|, |Tuple|, |Dictionary|, |Funcref|, |Job|, |Channel|, |Blob|, |Class|
and |object| types are not automatically converted.
```

Author: Thanks.  Updated the doc.

example, to add up all the numbers in a list: >
	:exe 'let sum = ' .. join(nrlist, '+')

1.4 Tuples ~
						*tuple* *Tuple* *Tuples*
A Tuple is an ordered sequence of items.  An item can be of any type.  Items

> can be accessed by their index number.  A Tuple is unmutable.

Reviewer 2: immutable?

example, to add up all the numbers in a list: >
	:exe 'let sum = ' .. join(nrlist, '+')

1.4 Tuples ~
						*tuple* *Tuple* *Tuples*
						*E1531* *E1532*
A Tuple is an ordered sequence of items.  An item can be of any type.  Items
can be accessed by their index number.  A Tuple is immutable.

Tuple creation ~
						*E1525* *E1526*

> A Tuple is created with a comma-separated list of items in parenthesis.

Reviewer: ```suggestion
A Tuple is created with a comma-separated sequence of items in parentheses.
```

Compare with these summaries above:

https://github.com/vim/vim/blob/cb50cda650e09eb009a0b11184cfd3db11c8a3dc/runtime/doc/eval.txt#L64-L74

More ‘comma-separated sequence’ candidates are at lines 275  
(276) (Lists):

https://github.com/vim/vim/blob/cb50cda650e09eb009a0b11184cfd3db11c8a3dc/runtime/doc/eval.txt#L275
 
and 714 (Dictionaries):

https://github.com/vim/vim/blob/cb50cda650e09eb009a0b11184cfd3db11c8a3dc/runtime/doc/eval.txt#L714

Author: I have updated the help text to incorporate this comment.

	:let last = mytuple[-1]		" get the last item: "four"

To avoid an error for an invalid index use the |get()| function.  When an item
is not available it returns zero or the default value you specify: >
	:echo get(mytuple, idx)
	:echo get(mytuple, idx, "NONE")

Tuple concatenation ~
							*tuple-concatenation*
Two tuples can be concatenated with the "+" operator: >
	:let longtuple = mytuple + (5, 6)
	:let longtuple = (5, 6) + mytuple
To prepend or append an item, turn it into a tuple by putting () around it.
The item must be followed by a comma.

Note that a tuple is immutable and items cannot be added or removed from a
tuple.

Subtuple ~
							*subtuple*
A part of the Tuple can be obtained by specifying the first and last index,
separated by a colon in square brackets: >
	:let shorttuple = mytuple[2:-1]	" get Tuple (3, "four")

Omitting the first index is similar to zero.  Omitting the last index is
similar to -1. >
	:let endtuple = mytuple[2:]	" from item 2 to the end: (3, "four")
	:let shorttuple = mytuple[2:2]	" Tuple with one item: (3)
	:let othertuple = mytuple[:]	" make a copy of the Tuple

Notice that the last index is inclusive.  If you prefer using an exclusive

> index use the |slice()| method.

Reviewer: ```suggestion
index use the |slice()| function.
```

Another `|slice()|` _method_ is referenced at line 335 (336).

Author: I have updated the help text to incorporate this comment.

If the first index is beyond the last item of the Tuple or the second item is
before the first item, the result is an empty tuple.  There is no error
message.

> If the second index is equal to or greater than the length of the tuple the

Reviewer: ```suggestion
If the second index is equal to or greater than the length of the tuple, the
```

Author: I have updated the help text to incorporate this comment.

length minus one is used: >
	:let mytuple = (0, 1, 2, 3)
	:echo mytuple[2:8]		" result: (2, 3)

NOTE: mytuple[s:e] means using the variable "s:e" as index.  Watch out for
using a single letter variable before the ":".  Insert a space when needed:
mytuple[s : e].

Tuple identity ~
							*tuple-identity*
When variable "aa" is a tuple and you assign it to another variable "bb", both
variables refer to the same tuple: >

> 	:let aa = [1, 2, 3]

Reviewer: ```suggestion
	:let aa = (1, 2, 3)
```

Author: I have updated the help text to incorporate this comment.

	:let bb = aa
<

Making a copy of a tuple is done with the |copy()| function.  Using [:] also
works, as explained above.  This creates a shallow copy of the tuple: Changing
a list item in the tuple will also change the item in the copied tuple: >
	:let aa = ([1, 'a'], 2, 3)
	:let bb = copy(aa)
	:let aa[0][1] = 'aaa'
	:echo aa
<	([1, aaa], 2, 3) >
	:echo bb
<	([1, aaa], 2, 3)

To make a completely independent tuple use |deepcopy()|.  This also makes a
copy of the values in the tuple, recursively.  Up to a hundred levels deep.

The operator "is" can be used to check if two variables refer to the same

> Tuple.  "isnot" does the opposite.  In contrast "==" compares if two tuples

Reviewer: ```suggestion
Tuple.  "isnot" does the opposite.  In contrast, "==" compares if two tuples
```

Author: I have updated the help text to incorporate this comment.

have the same value. >
	:let atuple = (1, 2, 3)
	:let btuple = (1, 2, 3)
	:echo atuple is btuple
<	0 >
	:echo atuple == btuple
<	1

Note about comparing tuples: Two tuples are considered equal if they have the
same length and all items compare equal, as with using "==".  There is one
exception: When comparing a number with a string they are considered
different.  There is no automatic type conversion, as with using "==" on
variables.  Example: >
	echo 4 == "4"
<	1 >

> 	echo (4) == ("4")

Reviewer: ```suggestion
	echo (4,) == ("4",)
```

Author: I have updated the help text to incorporate this comment.

<	0

Thus comparing Tuples is more strict than comparing numbers and strings.  You
can compare simple values this way too by putting them in a tuple: >

	:let a = 5
	:let b = "5"
	:echo a == b
<	1 >

> 	:echo (a) == (b)

Reviewer: ```suggestion
	:echo (a,) == (b,)
```

Author: I have updated the help text to incorporate this comment.

<	0

Tuple unpack ~

To unpack the items in a tuple to individual variables, put the variables in
square brackets, like list items: >
	:let [var1, var2] = mytuple

When the number of variables does not match the number of items in the tuple
this produces an error.  To handle any extra items from the tuple append ";"
and a variable name: >
	:let [var1, var2; rest] = mytuple

This works like: >
	:let var1 = mytuple[0]
	:let var2 = mytuple[1]
	:let rest = mytuple[2:]

Except that there is no error if there are only two items.  "rest" will be an
empty list then.

Tuple functions ~
						*E1535*
Functions that are useful with a Tuple: >
	:if empty(tuple)		" check if tuple is empty
	:let l = len(tuple)		" number of items in tuple
	:let big = max(tuple)		" maximum value in tuple
	:let small = min(tuple)		" minimum value in tuple

> 	:let xs = count(tuple, 'x')	" count nr of 'x' in tuple

Reviewer: ```suggestion
	:let xs = count(tuple, 'x')	" count number of 'x's in tuple
```

Author: I have updated the help text to incorporate this comment.

Tuple index ~
							*tuple-index* *E1518*
An item in the Tuple can be accessed by putting the index in square brackets
after the Tuple.  Indexes are zero-based, thus the first item has index zero. >
	:let item = mytuple[0]		" get the first item: 1
	:let item = mytuple[2]		" get the third item: 3

When the resulting item is a tuple this can be repeated: >
	:let item = nesttuple[0][1]	" get the first tuple, second item: 12
<
A negative index is counted from the end.  Index -1 refers to the last item in
the Tuple, -2 to the last but one item, etc. >
	:let last = mytuple[-1]		" get the last item: "four"

To avoid an error for an invalid index use the |get()| function.  When an item
is not available it returns zero or the default value you specify: >
	:echo get(mytuple, idx)
	:echo get(mytuple, idx, "NONE")

Tuple concatenation ~
							*tuple-concatenation*
Two tuples can be concatenated with the "+" operator: >
	:let longtuple = mytuple + (5, 6)
	:let longtuple = (5, 6) + mytuple
To prepend or append an item, turn it into a tuple by putting () around it.
The item must be followed by a comma.

							*E1539*
A variadic tuple cannot be concatenated with another tuple.

> 

Reviewer 3: This isn't very friendly. How to do the following?
```
type TNum = tuple<...list<number>>
var t1: TNum
var t2: TNum
var t3 = t1 + t2 ### or var t3: TNum = t1 + t2

# I guess do the following
var t3: TNum = list2tuple(tuple2list(t1) + tuple2list(t2))
```

How about limited restrictions. Don't think the following are ambiguous
- all items are the same type
- all items are automatically convertable
  ```
  tuple<...list<number>> + tuple<...list<float>> produces tuple<...list<float>>
  ```
- A binary operation where the LHS is not variadic and the RHS is variadic
- If the LHS of assignment is typed, should be able to determine at compile time if assignment is OK for arbitrary assignment RHS

It may not be practical to allow anything non-ambiguous, but the current restriction seems too severe.

Author: The latest PR relaxes the check for the variadic tuples in a concatenation operation.  The above cases are handled now.  Let me know if you see any issues with this.

example, to add up all the numbers in a list: >
	:exe 'let sum = ' .. join(nrlist, '+')

1.4 Tuples ~
						*tuple* *Tuple* *Tuples*
						*E1532* *E1533*
A Tuple is an ordered sequence of items.  An item can be of any type.  Items
can be accessed by their index number.  A Tuple is immutable.

A Tuple uses less memory compared to a List and provides O(1) lookup time.

Tuple creation ~
						*E1526* *E1527*
A Tuple is created with a comma-separated sequence of items in parentheses.
Examples: >
	:let mytuple = (1, "two", 3, "four")
	:let tuple = (5,)
	:let emptytuple = ()

An item can be any expression.  If there is only one item in the tuple, then

> the item must be followed by a comma.

Reviewer 4: I stumbled upon this.  I guess this is to distinguish this from simply grouping an expression: `(5)`

Author: Yes.  Without a comma, it will interpreted as grouping an expression and the result will be `5`.

Using a Tuple for an item creates a Tuple of Tuples: >
	:let nesttuple = ((11, 12), (21, 22), (31, 32))

Tuple index ~
							*tuple-index* *E1519*
An item in the Tuple can be accessed by putting the index in square brackets

> after the Tuple.  Indexes are zero-based, thus the first item has index zero. >

Reviewer: ```suggestion
after the Tuple.  Indexes are zero-based, thus the first item has index zero.
>
```

Author: Updated the doc.

When the resulting item is a tuple this can be repeated: >
	:let item = nesttuple[0][1]	" get the first tuple, second item: 12
<
A negative index is counted from the end.  Index -1 refers to the last item in
the Tuple, -2 to the last but one item, etc. >
	:let last = mytuple[-1]		" get the last item: "four"

To avoid an error for an invalid index use the |get()| function.  When an item
is not available it returns zero or the default value you specify: >
	:echo get(mytuple, idx)
	:echo get(mytuple, idx, "NONE")

Tuple concatenation ~
							*tuple-concatenation*
Two tuples can be concatenated with the "+" operator: >
	:let longtuple = mytuple + (5, 6)
	:let longtuple = (5, 6) + mytuple
To prepend or append an item, turn it into a tuple by putting () around it.
The item must be followed by a comma.

							*E1540*
Two variadic tuples with same item type can be concatenated but with different
item types cannot be concatenated.  Examples: >
    var a: tuple<...list<number>> = (1, 2)
    var b: tuple<...list<string>> = ('a', 'b')
    echo a + b		# not allowed

    var a: tuple<number, number> = (1, 2)
    var b: tuple<...list<string>> = ('a', 'b')

>     echo a + b		# allowed

Reviewer 4: This made me wonder. Why is this allowed, but this one isn't:
```
   echo a + b		# allowed
   echo b + a		# not allowed
```
is this because what matters is the first one being variadic?

Author: Yes.  If the type of the left operand is a variadic, then the types of both operands cannot be combined.

    var a: tuple<...list<number>> = (1, 2)
    var b: tuple<number, number> = (3, 4)
    echo a + b		# not allowed

    var a: tuple<...list<number>> = (1, 2)
    var b: tuple<number, ...list<number>> = (3, 4)
    echo a + b		# not allowed
<
Note that a tuple is immutable and items cannot be added or removed from a
tuple.

Subtuple ~
							*subtuple*
A part of the Tuple can be obtained by specifying the first and last index,
separated by a colon in square brackets: >
	:let shorttuple = mytuple[2:-1]	" get Tuple (3, "four")

Omitting the first index is similar to zero.  Omitting the last index is
similar to -1. >
	:let endtuple = mytuple[2:]	" from item 2 to the end: (3, "four")

> 	:let shorttuple = mytuple[2:2]	" Tuple with one item: (3)

Reviewer: ```suggestion
	:let shorttuple = mytuple[2:2]	" Tuple with one item: (3, )
```

Author: Updated the doc.

	:let othertuple = mytuple[:]	" make a copy of the Tuple

Notice that the last index is inclusive.  If you prefer using an exclusive

> index use the |slice()| function.

Reviewer: ```suggestion
index, use the |slice()| function.
```

Author: Updated the doc.

If the first index is beyond the last item of the Tuple or the second item is

> before the first item, the result is an empty tuple.  There is no error

Reviewer: ```suggestion
If the first index is beyond the last item of the Tuple or the last index is
before the first index, the result is an empty tuple.  There is no error
```

(This change and the following changes within the Subtuple   
section can also be applied to the Sublist section above.)

Author: The help text is referring to using negative indices with List and Tuples.  In this case, if the second index is before the first item, then the resulting value is empty: e.g.
```
echo [1, 2, 3][:-4]
echo (1, 2, 3)[:-4]
```

Author: Updated the doc to replace "second item" with "second index".

Reviewer: > The help text is referring to using negative indices with List and Tuples.

Speaking in behalf of first-time readers, I find it hard to  
penetrate the meaning. :)

Both ‘first’ and ‘last’ are described in the [first paragraph](https://github.com/vim/vim/blob/2a2cf4c9295ced4e6200bb52a79d3eb5fa4397b5/runtime/doc/eval.txt#L615-L626)  
of the Subtuple section in the context of sequence indexing  
`[first : last]`, and are also referenced in the following two  
paragraphs.  However, ‘second’ is substituted for ‘last’ in  
the [fourth and fifth paragraphs](https://github.com/vim/vim/blob/2a2cf4c9295ced4e6200bb52a79d3eb5fa4397b5/runtime/doc/eval.txt#L628-L635); while relevant examples  
indulge in _elements_ like `1`, `"two"` to boot:

```diff
(Next to *E696*.)
-	:let mylist = [1, two, 3, "four"]
+	:let mylist = [1, "two", 3, "four"]

(Next to *E1526*.)
	:let mytuple = (1, "two", 3, "four")

(Next to *E720*.)
	:let mydict = {1: 'one', 2: 'two', 3: 'three'}

(Next to "Accessing entries".)
+	:let mydict = {'one': 1, 'two': 2, 'three': 3}
	:let val = mydict["one"]
	:let mydict["four"] = 4
	:let val = mydict.one
	:let mydict.four = 4
```

Fixing typos in the examples and using another pair of names  
for indexing, disconnected from numeric references, e.g.  
‘start’ and ‘end’ after `:help slice()`, can improve their  
reading (also in the Sublist [section](https://github.com/vim/vim/blob/2a2cf4c9295ced4e6200bb52a79d3eb5fa4397b5/runtime/doc/eval.txt#L325-L345)).

Author: I have updated the help text to incorporate your comments.  Can you check the latest PR?

Reviewer: Thank you, it is good enough.

message.

> If the second index is equal to or greater than the length of the tuple, the

Reviewer: ```suggestion
If the last index is equal to or greater than the length of the tuple, the
```

Author: Only two indices can be specified for a slice.  So referring to it as "the second index"  makes sense to me.

length minus one is used: >
	:let mytuple = (0, 1, 2, 3)
	:echo mytuple[2:8]		" result: (2, 3)

NOTE: mytuple[s:e] means using the variable "s:e" as index.  Watch out for
using a single letter variable before the ":".  Insert a space when needed:
mytuple[s : e].

Tuple identity ~
							*tuple-identity*
When variable "aa" is a tuple and you assign it to another variable "bb", both
variables refer to the same tuple: >
	:let aa = (1, 2, 3)
	:let bb = aa
<

Making a copy of a tuple is done with the |copy()| function.  Using [:] also

> works, as explained above.  This creates a shallow copy of the tuple: Changing
> a list item in the tuple will also change the item in the copied tuple: >

Reviewer: ```suggestion
works, as explained above.  This creates a shallow copy of the tuple.  For
example, changing a list item in the tuple will also change the item in the
copied tuple: >
```

Author: Updated the doc.

	:let aa = ([1, 'a'], 2, 3)
	:let bb = copy(aa)
	:let aa[0][1] = 'aaa'
	:echo aa
<	([1, aaa], 2, 3) >
	:echo bb
<	([1, aaa], 2, 3)

> To make a completely independent tuple use |deepcopy()|.  This also makes a

Reviewer: ```suggestion
To make a completely independent tuple, use |deepcopy()|.  This also makes a
```

Author: Updated the doc.

copy of the values in the tuple, recursively.  Up to a hundred levels deep.

The operator "is" can be used to check if two variables refer to the same
Tuple.  "isnot" does the opposite.  In contrast, "==" compares if two tuples
have the same value. >
	:let atuple = (1, 2, 3)
	:let btuple = (1, 2, 3)
	:echo atuple is btuple
<	0 >
	:echo atuple == btuple
<	1

Note about comparing tuples: Two tuples are considered equal if they have the
same length and all items compare equal, as with using "==".  There is one
exception: When comparing a number with a string they are considered
different.  There is no automatic type conversion, as with using "==" on
variables.  Example: >
	echo 4 == "4"
<	1 >
	echo (4,) == ("4",)
<	0

Thus comparing Tuples is more strict than comparing numbers and strings.  You
can compare simple values this way too by putting them in a tuple: >

	:let a = 5
	:let b = "5"
	:echo a == b
<	1 >
	:echo (a,) == (b,)
<	0

Tuple unpack ~

To unpack the items in a tuple to individual variables, put the variables in
square brackets, like list items: >
	:let [var1, var2] = mytuple

When the number of variables does not match the number of items in the tuple

> this produces an error.  To handle any extra items from the tuple append ";"

Reviewer: ```suggestion
this produces an error.  To handle any extra items from the tuple, append ";"
```

Author: Updated the doc.

> and a variable name: >

Reviewer 4: ```suggestion
and a variable name (which will then be of type tuple/list)?: >
```

Author: The type will be "tuple".

	:let [var1, var2; rest] = mytuple

This works like: >
	:let var1 = mytuple[0]
	:let var2 = mytuple[1]
	:let rest = mytuple[2:]

Except that there is no error if there are only two items.  "rest" will be an

> empty list then.

Reviewer 4: should that be an empty `list`? or empty `tuple`?

Author: It should be empty "tuple".  Updated the code to return a tuple and the doc.

	+ expr9			unary plus

|expr10|  expr11
	expr10[expr1]		byte of a String or item of a |List| or
				|Tuple|
	expr10[expr1 : expr1]	substring of a String or sublist of a |List|

> 				or |Tuple|

Reviewer: ```suggestion
				or subtuple of a |Tuple|
```

Author: Updated the doc.

	if get(Part1, 'name') == get(Part2, 'name')
	   " Part1 and Part2 refer to the same function
<							*E1037*
Using "is" or "isnot" with a |List|, |Tuple|, |Dictionary| or |Blob| checks
whether the expressions are referring to the same |List|, |Tuple|,

> |Dictionary| or |Blob| instance.  A copy of a |List|  or |Tuple| is different

Reviewer: ```suggestion
|Dictionary| or |Blob| instance.  A copy of a |List| or |Tuple| is different
```

Author: Updated the doc.

from the original |List| or |Tuple|.  When using "is" without a |List|,
|Tuple|, |Dictionary| or |Blob|, it is equivalent to using "equal", using

> "isnot" equivalent to using "not equal".  Except that a different type means
> the values are different: >

Reviewer: ```suggestion
"isnot" is equivalent to using "not equal".  Except that a different type
means the values are different: >
```

Author: Updated the doc.

expr6 and expr7				*expr6* *expr7* *E1036* *E1051*
---------------
								*expr-+*
expr7 + expr7   Number addition, |List| or |Tuple| or |Blob| concatenation
expr7 - expr7   Number subtraction				*expr--*
expr7 . expr7   String concatenation				*expr-.*
expr7 .. expr7  String concatenation				*expr-..*

For |Lists| only "+" is possible and then both expr7 must be a list.  The

> result is a new list with the two lists Concatenated.  Same for a |Tuple|.

Reviewer: ```suggestion
result is a new list with the two lists concatenated.  Same for a |Tuple|.
```

Author: Updated the doc.

			|List| item.

:let [{name}, ..., ; {lastname}] = {expr1}				*E452*
			Like |:let-unpack| above, but the |List| or |Tuple|
			may have more items than there are names.  A list of
			the remaining items is assigned to {lastname}.  If

> 			there are no remaining items {lastname} is set to an

Reviewer: ```suggestion
			there are no remaining items, {lastname} is set to an
```

Author: Updated the doc.

			|List| item.

:let [{name}, ..., ; {lastname}] = {expr1}				*E452*
			Like |:let-unpack| above, but the |List| or |Tuple|
			may have more items than there are names.  A list of
			the remaining items is assigned to {lastname}.  If
			there are no remaining items, {lastname} is set to an
			empty list.
			Example: >

> 				:let [a, b; rest] = ["aval", "bval", 3, 4]

Reviewer: The unpacking entry should mention both `{lastname}` lists *and  
tuples* to align it with:

https://github.com/vim/vim/blob/a0c917ea9809281ada9906c91c77b9b92d2e0f11/runtime/doc/eval.txt#L409-L426

and

https://github.com/vim/vim/blob/a0c917ea9809281ada9906c91c77b9b92d2e0f11/runtime/doc/eval.txt#L696-L713

And a tuple example added.

Author: Updated the doc to mention a tuple and added an example.

			get an error message: "E940: Cannot lock or unlock
			variable {name}".

> 			[depth] is relevant when locking a |List| or |Tuple| or
> 			|Dictionary|.  It specifies how deep the locking goes:

Reviewer: ```suggestion
			[depth] is relevant when locking a |List| or |Tuple|
			or |Dictionary|.  It specifies how deep the locking
			goes:
```

Author: Updated the doc.

				0	Lock the variable {name} but not its
					value.
				1	Lock the |List| or |Tuple| or
					|Dictionary| itself, cannot add or
					remove items, but can still change
					their values.
				2	Also lock the values, cannot change
					the items.  If an item is a |List| or
					|Tuple| or |Dictionary|, cannot add or
					remove items, but can still change the
					values.
				3	Like 2 but for the |List| / |Tuple| /
					|Dictionary| in the |List| / |Tuple| /
					|Dictionary|, one level deeper.
			The default [depth] is 2, thus when {name} is a |List|

> 			or |Tuple| or |Dictionary| the values cannot be changed.

Reviewer: ```suggestion
			or |Tuple| or |Dictionary| the values cannot be
			changed.
```

Author: Updated the doc.
